updated health docs for 21.6 (#34)
CiscoTestAutomation/genie: docs/health/health_design/design.rst

Design
======

> The previous page showed you how to use pyATS Health Check easily. This page is for those who want to use their own pyATS Health Check. If you are happy with previous page way, please move on :ref:`processor key in YAML<processor_key_in_yaml>`.

Reviewer: ```suggestion
The previous page showed you how to use pyATS Health Check easily. This page is for those who want to use their own pyATS Health Check. If you are happy with the way it was done on the previous page, then please move on :ref:`processor key in YAML<processor_key_in_yaml>`.
```

> Collect and monitor the state of your devices as your testscript is executing with pyATS Health Check. It can collect traceback, core files, etc. pyATS Health Check can be even yaml driven and it is based on :ref:`Blitz<blitz>`. All Blitz functionalities are fully supported in health check.

Reviewer 2: ```suggestion
You can collect and monitor the state of your devices as your testscript is executing with pyATS Health Check. It can collect traceback, core files, etc. pyATS Health Check is yaml driven and it is based on :ref:`Blitz<blitz>`. All Blitz functionalities are supported in Health Check.
```

> The health check is driven by a `health_file` which is provided at run time. There is two different mechanism to run the checks:

Reviewer: ```suggestion
The health check is driven by a `health_file` which is provided at run time. There are two different mechanism to run the checks:
```

1. Part of a pre/post processor of a section or testcase.

> 2. Part of another process continuously (Coming with 21.7).

Reviewer 2: ```suggestion
2. Continuous data collection using a background process (Coming with 21.7).
```

> To use own pyATS Health Checks, pyATS Health Check leverages existing :ref:`Blitz<blitz>` style YAML format and use same format and capabilities what Blitz has.

Reviewer 2: ```suggestion
To use your own pyATS Health Checks, you can specify pyATS Health Checks by leveraging the existing :ref:`Blitz<blitz>` style YAML format as it uses the same format and capabilities that Blitz has.
```

> The actions in pyATS Health Check yaml, those actions will be added to pyATS job's each section as pre/post processors. Pre, Post or both processors can be specified in the yaml.

Reviewer 2: ```suggestion
The actions in pyATS Health Check yaml will be added to pyATS job's sections as pre/post processors. Execution of the actions as pre, post or both can be specified in the yaml.
```

.. note::

     See the detail of :ref:`Blitz<blitz>`

.. code-block:: bash

    # Quick start Guide!
    # ------------------
    # 1. Create the health yaml

>     # 2. Add the arguments

Reviewer 2: add the arguments ... where? in the yaml file? to the job command line?

Author: yes. to pyats command. updated to be clear via 165442b

>     # 3. (Optional) Run it and hope no issue is found by the pyATS Health Check!

Reviewer 2: ```suggestion
    # 3. Run the job and verify issues found by the pyATS Health Check if any
```

    # Integrated with pyATS jobs
    pyats run job <job file> --testbed-file /path/to/testbed.yaml --health-file /path/to/health.yaml

.. note::

>      Devices for pyATS Health Check need to be in testbed yaml for pyATS job

Reviewer: ```suggestion
     Devices for pyATS Health Check need to be in the testbed yaml for the pyATS job
```

>      Devices for pyATS Health Check need to be in the testbed yaml for the pyATS job

Reviewer 2: ```suggestion
     Devices for pyATS Health Check need to be specified in the health yaml file and the testbed yaml for the pyATS job to execute health checks against the device.
```

     The pyATS testbed object will be converted to Genie testbed object to have Genie functionalities.

--------

.. figure:: health_run.png
    :width: 350
    :align: right

> Create health.yaml with knowledge about pyATS Blitz and then just add argument `--health-file` with the file. pyATS Health Check will add Blitz actions as processors to sections in pyATS job. And pyATS Health Check processors will run before and after each section to monitor/collect device status and etc.

Reviewer 2: ```suggestion
Create health.yaml with knowledge about pyATS Blitz and then add the argument `--health-file` with the filename. pyATS Health Check will add Blitz actions as processors to sections in pyATS job. pyATS Health Check processors will run before and after each section (as specified) to monitor/collect device status as specified in the health yaml file.
```

> Even for development of health yaml, the health file makes developer's life easier. It's same format with Blitz format and the health file can run like Blitz as well. So, develop and test as Blitz first and just switch argument from `--trigger-datafile` to `--health-file` when health yaml is ready.

Reviewer: ```suggestion
Even for development of health yaml, the health file makes developer's life easier. It's the same format as the Blitz format and the health file can run like Blitz as well. So, develop and test as Blitz first and just switch the argument from `--trigger-datafile` to `--health-file` when health yaml is ready.
```

pyATS Health Check comes with default checks; however it is fully open-sourced! You can add your own checks to be executed at any time! Any features or functions to monitor, those can be contributed and developed as Blitz actions or APIs.

.. note::

     See all the available APIs for pyATS Health Check `Available Apis <https://pubhub.devnetcloud.com/media/genie-feature-browser/docs/#/apis>`_

--------

> pyATS Health Check's added processors can be easily found in pyATS Logviewer or XPRESSO. The added processors has `pyATS Health Check` and also icons for both `Pre` and `Post`.

Reviewer 2: ```suggestion
pyATS Health Check's added processors can be easily found in pyATS or XPRESSO logviewer. The processor results are added with a `pyATS Health Check` label and also icons for both `Pre` and `Post`.
```

      source:
        pkg: genie.libs.health
        class: health.Health

>       devices: ['uut']

Reviewer 2: Is this no longer needed? See my comment above, my understanding was the devices need to in both health.yaml and testbed.yaml

Author: ok to have `device` under `action` instead in case of Blitz

                processor: both
                # `function` can be found from Genie Feature Browser
                # Please find the link to the page from bottom of this section

>                 function: health_cpu

Reviewer 2: why is there a new health specific API?

Author: same with https://github.com/CiscoTestAutomation/genie/pull/34#discussion_r666936546

                arguments:
                  command: show processes cpu
                  processes: ['BGP I/O']
        - memory:
            - api:
                device: uut
                processor: post

>                 function: health_memory

Reviewer 2: why is there a health specific api?

Author: good catch! yes. dedicated health API starting with `health_`. but it returns more data for health. but it doesn't mean user cannot use generic APIs. user still can use generic genie APIs as well if they want.

                arguments:
                  command: show processes memory
                  processes: ['\*Init\*']
                include:
                  - sum_value_operator('value', '<', 90)

.. note::

  All available APIs and Parsers can be found here `Genie Feature Browser <https://pubhub.devnetcloud.com/media/genie-feature-browser/docs/#/>`_

>   Make sure you read the comments above! After this all will make sense!

Reviewer 2: I think we can remove this line...

Author: removed via 92ea7ca

There are three ways to filter it. By default pyATS Health Checks run before and after every testcase and section. With the filtering you can decide where they are executed.

> Testcase level: health-tc-uids, provide the testcase/trigger names from Testcase/Trigger datafile. The exact name can be provided or regular expression is also supported. pyATS Health Check processors will run only for the given testcase/trigger names which match the full name or match the regex.

Reviewer 2: why has `-tc` been added? 'common_setup' / 'CommonSetup' is technically not a testcase but it can still have a processor attached to it

Author: It was requested to change internally. still old name `health-sections`/`health-uids` works as deprecated. I will take a note to discuss. `-tc` might be confusing. but no functionality issue even against `common_setup/`CommonSetup`
